agentendpoint/exec_task.go

```diff
@@ -16,14 +16,91 @@ package agentendpoint
 
 import (
 	"context"
+	"errors"
 	"fmt"
+	"io"
+	"os"
+	"os/exec"
+	"path"
+	"path/filepath"
+	"runtime"
 	"time"
 
+	"cloud.google.com/go/storage"
+	"github.com/GoogleCloudPlatform/guest-logging-go/logger"
 	"github.com/GoogleCloudPlatform/osconfig/config"
+	"github.com/GoogleCloudPlatform/osconfig/external"
 
 	agentendpointpb "github.com/GoogleCloudPlatform/osconfig/_internal/gapi-cloud-osconfig-go/google.golang.org/genproto/googleapis/cloud/osconfig/agentendpoint/v1alpha1"
 )
 
+var (
+	winRoot = os.Getenv("SystemRoot")
+	sh      = "/bin/sh"
+
+	winPowershell string
+	winCmd        string
+
+	goos = runtime.GOOS
+
+	errLinuxPowerShell = errors.New("interpreter POWERSHELL cannot be used on non-Windows system")
+	errWinNoInt        = fmt.Errorf("interpreter must be specified for a Windows system")
+)
+
+func init() {
+	if winRoot == "" {
+		winRoot = `C:\Windows`
+	}
+	winPowershell = filepath.Join(winRoot, `System32\WindowsPowerShell\v1.0\PowerShell.exe`)
+	winCmd = filepath.Join(winRoot, `System32\cmd.exe`)
+}
+
+var run = func(cmd *exec.Cmd) ([]byte, error) {
+	return cmd.CombinedOutput()
+}
+
+func getGCSObject(ctx context.Context, gcsObject *agentendpointpb.GcsObject, loggingLables map[string]string) (string, error) {
+	if gcsObject == nil {
+		return "", errors.New("gcsObject cannot be nil")
+	}
+
+	var reader io.ReadCloser
+	cl, err := storage.NewClient(ctx)
+	if err != nil {
+		return "", fmt.Errorf("error creating gcs client: %v", err)
+	}
+	external.FetchGCSObject(ctx, cl, gcsObject.Object, gcsObject.Bucket, gcsObject.GenerationNumber)
+	defer reader.Close()
```

Comment:
reader needs to be assigned, see https://github.com/GoogleCloudPlatform/osconfig/commit/10ef0dbda51d24ea5f2d927d9ea65182f7d0851a, it was dropped after the package common -> package external refactor